Fixes #21536: Allowing the ability to pass aws session paramters to credstash lookup

beneshed opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
#21536
in credstash.getSecret calls get_session(**kwargs) which is looking for 
```
def get_session(aws_access_key_id=None, aws_secret_access_key=None,
                aws_session_token=None, profile_name=None):
```
So I extracted the arguments out of the **kwargs if they exist, pass them into kwargs_pass and add it to credstash.getSecret
<!---
If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Feature Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
plugins/lookup/credstash

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```
ansible 2.2.2.0
```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```

```

Commit:
Allowing the ability to pass aws session paramters to credstash lookup
Fixes #21536
https://github.com/ansible/ansible/issues/21536

willthames commented:
This looks good - I didn't even know about the credstash plugin before now.

I can definitely see the usecase for this change - being able to use either a profile or the results of sts_assume_role both make sense.

Do you know how this behaves when environment variables are set? Do the environment variables work? Do the kwargs take precedence over the environment variables (which is what I'd expect)

willthames commented:
There is a large amount of overlap with #23003 - I'd recommend merging #23003 first and then this one to get the extra key variables.

thebenwaters commented:
@willthames the way the plugin was originally written, I believe that there's no way for it to access the environment variables. It calls getSecret, which calls get_session https://github.com/fugue/credstash/blob/master/credstash.py#L435. I can look for environment variables and if not set just pass it down a chain?

thebenwaters committed:
Default environment variable lookup

To get around non lookup of environment variables. First, check for environment variables, then look for parameters

thebenwaters commented:
The only issue with this is, if you have an environment variable set unintentionally and manually set a parameter in the lookup. It will not take the parameter. Thoughts?

thebenwaters committed:
Following AWS Precedence

http://docs.aws.amazon.com/cli/latest/userguide/cli-chap-getting-started.html#config-settings-and-precedence

thebenwaters commented:
Switched to awscli order of precedence

http://docs.aws.amazon.com/cli/latest/userguide/cli-chap-getting-started.html#config-settings-and-precedence

willthames commented:
shipit, thanks for the precedence updates, definitely a worthwhile fix.
